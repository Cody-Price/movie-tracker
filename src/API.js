import cleanMovies from './helper'

export const fetchMovies = async (url) => {
  try {
    const response = await fetch(url)
    const result = await response.json()
    return cleanMovies(result.results)
  } catch(err) {
    console.log(err.message)
  }
}

export const createUser = async (user) => {
  const url = 'http://localhost:3000/api/users/new'
  const allUsers = await fetchAllUsers()
  const matchingUser = allUsers.find(currUser => currUser.email === user.email )
  if (matchingUser) {
    return matchingUser
  } else {
      try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(user)
      })
       
    } catch(err) {
      console.log(err.message);
    }
  }
}

export const loginUser = async (loginUser) => {
  const allUsers = await fetchAllUsers()
  const matchedUser = allUsers.find(user => {
        return user.email === loginUser.email && user.password === loginUser.password;
      })
  return matchedUser;
}

export const fetchAllUsers = async () => {
  const url = 'http://localhost:3000/api/users/';
  try {
    const response = await fetch(url)
    const data = await response.json()
    return data.data
  } catch(err) {
    console.log(err.message);
  }
}