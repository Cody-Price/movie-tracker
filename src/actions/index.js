export const loginUser = (userObject) => {
  return {
    type: 'LOGIN_USER',
    userObject
  }
}

export const logout = () => {
  return {
    type: 'LOGOUT_USER'
  }
}

export const addMovies = (movies) => {
  return {
    type: 'ADD_MOVIES',
    movies
  }
}

export const addFavorite = (favorite) => {
  return {
    type: 'ADD_FAVORITE',
    favorite
  }
}