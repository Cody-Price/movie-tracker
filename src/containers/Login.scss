.hide-error {
  display: none
}

.login-container {
  align-items: center;
  background-color: #ba0000;
  border-radius: 20px;
  border: 2px solid white;
  color: white;
  display: flex;
  flex-direction: column;
  font-family: Arial, Helvetica, sans-serif;
  font-size: 20px;
  font-weight: bold;
  height: 350px;
  justify-content: center;
  margin: 0 auto;
  position: relative;
  text-shadow: 2px 2px black;
  top: 250px;
  width: 45%;
}

.login-form {
  position: relative;
  color: pink;
  margin: 30px;
  top: -20px;
  width: 100%;
}

input {
  border-radius: 10px;
  border: 2px solid black;
  margin: 10px;
  outline: none;
  padding: 5px 5px;
  transition-duration: 0.3s;
  width: 130px;

  &:focus {
    transform: scale(1.1);
    transition-duration: 0.3s;
  }
}

.signup-prompt {

}

label {
  margin: 0 16px;
  font-size: 16px;
}

.login-submit-btn {
  margin-left: 15px;
}