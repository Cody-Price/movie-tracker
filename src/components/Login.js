import React, { Component } from 'react';
import * as API from '../API';
import './Login.scss';

class Login extends Component {
  constructor(props) {
    super(props)
    console.log(props.history);
    this.state = {
      email: '',
      password: '',
      verifiedPassword: '',
      newUser: false,
      name: '',
      matchedUser: '',
      errorMessage: false
    }
  }

  displayNewUserForm = () => {
    this.resetState()
    this.setState({
      newUser: true
    })
  }

  handleChange = (e) => {
    const {name, value} = e.target
    this.setState({
      [name]: value
    })
  }

  handleNewUserSubmit = async (e) => {
    e.preventDefault()
    const email = this.state.email.toLowerCase();
    const {name, password} = this.state;
    const user = {name, password, email}
    const matchingUser = await API.createUser(user)
    console.log(matchingUser)
    if (matchingUser) {
      this.setState({
        errorMessage: true
      })
    } else {
      this.resetState();
      this.props.history.push('/');
    }
  }

  handleLoginSubmit = async (e) => {
    e.preventDefault()
    const email = this.state.email.toLowerCase();
    const {password} = this.state;
    const user = {password, email}
    const matchedUser = await API.loginUser(user)
    if (matchedUser === undefined) {
      this.setState({errorMessage: true})
    } else {
      this.setState({matchedUser})
      this.resetState()
      this.props.history.push('/');
    }
  }

  resetState = () => {
    this.setState({
      email: '',
      password: '',
      verifiedPassword: '',
      name: ''
    })
  }



  render() {
    const { email, password, verifiedPassword, name } = this.state;
    if (!this.state.newUser) {
      return (
        <section>
          <div className={this.state.errorMessage ? 'show-error' : 'hide-error'}>E-Mail and Password do not match</div>
          <form onSubmit={this.handleLoginSubmit}>
            <input
              type="text"
              placeholder="E-Mail"
              value={email}
              name="email"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="Password"
              value={password}
              name="password"
              onChange={this.handleChange}
            />
            <button>Submit</button>
          </form>
          <label>New to Movie Tracker?</label>
          <button onClick={this.displayNewUserForm}>Sign Up!</button>
        </section>
      )
    } else {
      return (
        <section>
          <div className={this.state.errorMessage ? 'show-error' : 'hide-error'}>E-Mail already exists</div>
          <form onSubmit={this.handleNewUserSubmit}>
            <input
              type="text"
              placeholder="Name"
              value={name}
              name="name"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="E-Mail"
              value={email}
              name="email"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="Password"
              value={password}
              name="password"
              onChange={this.handleChange}
            />
            <button>Submit</button>
          </form>
        </section>
      )
    }
  }
}


export default Login;