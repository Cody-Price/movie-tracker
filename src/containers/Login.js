import React, { Component } from 'react';
import * as API from '../API';
import './Login.scss';
import { loginUser, addFavorites } from '../actions';
import { connect } from 'react-redux';

class Login extends Component {
  constructor(props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      verifiedPassword: '',
      newUser: false,
      name: '',
      errorMessage: false
    }
  }

  displayNewUserForm = () => {
    this.resetState()
    this.setState({
      newUser: true
    })
  }

  handleChange = (e) => {
    const {name, value} = e.target
    this.setState({
      [name]: value
    })
  }

  handleNewUserSubmit = async (e) => {
    e.preventDefault()
    const { name, password, email } = this.state;
    const user = {name, email, password}
    const userValid = await API.createUser(user)
    if (!userValid) {
      this.setState({
        errorMessage: true
      })
    } else {
      this.props.login({name: user.name, id: userValid})
      this.resetState();
      this.props.history.push('/');
    }
  }

  handleLoginSubmit = async (e) => {
    e.preventDefault()
    const email = this.state.email.toLowerCase();
    const {password} = this.state;
    const user = {password, email}
    const userObject = await API.loginUser(user)
    if (!userObject) {
      this.setState({errorMessage: true})
    } else {
      this.props.login({name: userObject.name, id: userObject.id});
      const favorites = await API.fetchFavorites(userObject.id);
      this.props.addFavorites(favorites);
      this.resetState();
      this.props.history.push('/');
    }
  }

  resetState = () => {
    this.setState({
      email: '',
      password: '',
      verifiedPassword: '',
      name: ''
    })
  }

  render() {
    const { email, password, verifiedPassword, name } = this.state;
    if (!this.state.newUser) {
      return (
        <section className="login-container">
          <h2>Welcome back!</h2>
          <div className={this.state.errorMessage ? 'show-error' : 'hide-error'}>E-Mail and Password do not match</div>
          <form onSubmit={this.handleLoginSubmit} className="login-form">
            <input
              type="email"
              placeholder="E-Mail"
              value={email}
              name="email"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="Password"
              value={password}
              name="password"
              onChange={this.handleChange}
            />
            <button className="login-submit-btn">Submit</button>
          </form>
          <div className="signup-prompt">
            <label>Or are you new to Movie Tracker?</label>
            <button onClick={this.displayNewUserForm}>Sign Up!</button>
          </div>
        </section>
      )
    } else {
      return (
        <section className="login-container">
          <h2>Get ready to become a professional TRACKER!</h2>
          <div className={this.state.errorMessage ? 'show-error' : 'hide-error'}>E-Mail already exists</div>
          <form onSubmit={this.handleNewUserSubmit} className="login-form">
            <input
              type="text"
              placeholder="Name"
              value={name}
              name="name"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="E-Mail"
              value={email}
              name="email"
              onChange={this.handleChange}
            />
            <input
              type="text"
              placeholder="Password"
              value={password}
              name="password"
              onChange={this.handleChange}
            />
            <button className="login-submit-btn">Submit</button>
          </form>
        </section>
      )
    }
  }
}

const mapDispatchToProps = (dispatch) => ({
  login: (username) => dispatch(loginUser(username)),
  addFavorites: (favorites) => dispatch(addFavorites(favorites)) 
})

export default connect(null, mapDispatchToProps)(Login)