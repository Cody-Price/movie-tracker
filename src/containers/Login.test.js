import React from 'react';
import { shallow } from 'enzyme';
import { Login, mapDispatchToProps } from './Login'
import * as API from '../API';
import * as actions from '../actions'

describe('Login', () => {
  let mockFavorites
  let mockUser
  let wrapper

  beforeEach(() => {
    mockFavorites = [{title: 'movie1', movie_id: 1}, {title: 'movie2', movie_id: 2}]
    mockUser = { name: 'Taylor', id: 1 }
    wrapper = shallow(
      <Login
        login = { jest.fn() }
        addFavorites = { jest.fn() }
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should set newUser to true if sign up is clicked', () => {
    const initialState = {
      email: '',
      password: '',
      verifiedPassword: '',
      newUser: false,
      name: '',
      errorMessage: false
    }

    wrapper.setState(initialState)
    const button = wrapper.find('button').at(1).simulate('click')
    expect(wrapper.state().newUser).toEqual(true);  
  })

  it('should set reset state if sign up is clicked', () => {
    const spy = jest.spyOn(wrapper.instance(), 'resetState')
    const button = wrapper.find('button').at(1).simulate('click')
    expect(spy).toHaveBeenCalled()
  })



  describe('mapDispatchToProps', () => {
    let mockDispatch

    beforeEach(() => {
      mockDispatch = jest.fn()
    })
    
    it('should call dispatch with the correct params for login', () => {
      const user = { name: 'Cole' }
      const actionToDispatch = actions.loginUser(user)
      const result = mapDispatchToProps(mockDispatch)
      result.login(user)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch with the correct params for addFavorites', () => {
      const favorites = [{title: 'movie1', movie_id: 1}, {title: 'movie2', movie_id: 2}]
      const actionToDispatch = actions.addFavorites(favorites)
      const result = mapDispatchToProps(mockDispatch)
      result.addFavorites(favorites)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})