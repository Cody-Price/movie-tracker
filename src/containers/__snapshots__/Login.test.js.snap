// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    createUser={[MockFunction]}
    error="error"
    fetchFavorites={[MockFunction]}
    loginUser={[MockFunction]}
    user={
      Object {
        "id": 1,
        "name": "Kylie",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Welcome back!
        </h2>,
        <div
          className="hide-error"
        >
          E-Mail and Password do not match
        </div>,
        <form
          className="login-form"
          onSubmit={[Function]}
        >
          <input
            name="email"
            onChange={[Function]}
            placeholder="E-Mail"
            required={true}
            type="email"
            value=""
          />
          <input
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
          />
          <button
            className="login-submit-btn"
          >
            Submit
          </button>
        </form>,
        <div
          className="signup-prompt"
        >
          <label>
            Or are you new to Movie Tracker?
          </label>
          <button
            onClick={[Function]}
          >
            Sign Up!
          </button>
        </div>,
      ],
      "className": "login-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Welcome back!",
        },
        "ref": null,
        "rendered": "Welcome back!",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "E-Mail and Password do not match",
          "className": "hide-error",
        },
        "ref": null,
        "rendered": "E-Mail and Password do not match",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-Mail"
              required={true}
              type="email"
              value=""
            />,
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value=""
            />,
            <button
              className="login-submit-btn"
            >
              Submit
            </button>,
          ],
          "className": "login-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "E-Mail",
              "required": true,
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "required": true,
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "className": "login-submit-btn",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Or are you new to Movie Tracker?
            </label>,
            <button
              onClick={[Function]}
            >
              Sign Up!
            </button>,
          ],
          "className": "signup-prompt",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Or are you new to Movie Tracker?",
            },
            "ref": null,
            "rendered": "Or are you new to Movie Tracker?",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up!",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Sign Up!",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Welcome back!
          </h2>,
          <div
            className="hide-error"
          >
            E-Mail and Password do not match
          </div>,
          <form
            className="login-form"
            onSubmit={[Function]}
          >
            <input
              name="email"
              onChange={[Function]}
              placeholder="E-Mail"
              required={true}
              type="email"
              value=""
            />
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value=""
            />
            <button
              className="login-submit-btn"
            >
              Submit
            </button>
          </form>,
          <div
            className="signup-prompt"
          >
            <label>
              Or are you new to Movie Tracker?
            </label>
            <button
              onClick={[Function]}
            >
              Sign Up!
            </button>
          </div>,
        ],
        "className": "login-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Welcome back!",
          },
          "ref": null,
          "rendered": "Welcome back!",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "E-Mail and Password do not match",
            "className": "hide-error",
          },
          "ref": null,
          "rendered": "E-Mail and Password do not match",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="email"
                onChange={[Function]}
                placeholder="E-Mail"
                required={true}
                type="email"
                value=""
              />,
              <input
                name="password"
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />,
              <button
                className="login-submit-btn"
              >
                Submit
              </button>,
            ],
            "className": "login-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "E-Mail",
                "required": true,
                "type": "email",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "required": true,
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "className": "login-submit-btn",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                Or are you new to Movie Tracker?
              </label>,
              <button
                onClick={[Function]}
              >
                Sign Up!
              </button>,
            ],
            "className": "signup-prompt",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Or are you new to Movie Tracker?",
              },
              "ref": null,
              "rendered": "Or are you new to Movie Tracker?",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sign Up!",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Sign Up!",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
